ocp_utilities/monitoring.py

```diff
@@ -0,0 +1,209 @@
+import json
+import re
+from json import JSONDecodeError
+
+import requests
+from ocp_resources.route import Route
+from ocp_resources.secret import Secret
+from ocp_resources.service_account import ServiceAccount
+from ocp_resources.utils import TimeoutExpiredError, TimeoutSampler
+from simple_logger.logger import get_logger
+
+from ocp_utilities.infra import get_client
+
+
+TIMEOUT_2MIN = 2 * 60
+TIMEOUT_10MIN = 10 * 60
+
+LOGGER = get_logger(name=__name__)
+
+
+class Prometheus(object):
+    """
+    For accessing Prometheus cluster metrics
+
+    Prometheus HTTP API doc:
+    https://prometheus.io/docs/prometheus/latest/querying/api/
+
+    Argument for query method should be the entire string following the server address
+        e.g.
+        prometheus = Prometheus()
+        up = prometheus.query("/api/v1/query?query=up")
+    """
+
+    def __init__(
+        self, namespace=None, resource_name=None, client=None, verify_ssl=True
+    ):
+        self.namespace = namespace or "openshift-monitoring"
+        self.resource_name = resource_name or "prometheus-k8s"
+        self.client = client or get_client()
+        self.api_v1 = "/api/v1"
+        self.verify_ssl = verify_ssl
+
+        route = Route(
+            namespace=self.namespace, name=self.resource_name, client=self.client
+        ).instance.spec.host
+
+        self.api_url = f"https://{route}"
+        self.headers = self._get_headers()
+        self.scrape_interval = self.get_scrape_interval() or 30
+
+    def _get_headers(self):
+        """Uses the Prometheus serviceaccount to get an access token for OAuth"""
+
+        LOGGER.info("Setting Prometheus headers and Obtaining OAuth token")
+
+        prometheus_sa = ServiceAccount(
+            namespace=self.namespace, name=self.resource_name, client=self.client
+        )
+        secret = Secret(
+            namespace=self.namespace,
+            name=prometheus_sa.instance.imagePullSecrets[0].name,
+            client=self.client,
+        )
+
+        token = secret.instance.metadata.annotations["openshift.io/token-secret.value"]
+
+        return {"Authorization": f"Bearer {token}"}
+
+    def _get_response(self, query):
+        response = requests.get(
+            f"{self.api_url}{query}", headers=self.headers, verify=self.verify_ssl
+        )
+
+        try:
+            return json.loads(response.content)
+        except JSONDecodeError as json_exception:
+            LOGGER.error(
+                "Exception converting query response to JSON: "
+                f"exc={json_exception} response_status_code={response.status_code} response={response.content}"
+            )
+            raise
+
+    def query(self, query):
+        """get the prometheus query result
+
+        Args:
+            query(str): promthetheus query string
+
+        Returns:
+            dict: query result
+        """
+        return self._get_response(query=f"{self.api_v1}/query?query={query}")
+
+    def get_alert(self, alert_name):
```

Comment:
Is this function going to return one alert or multiple alerts for a given alert name?
If multiple then the function name need to be renamed.

Reply:
it will return multiple alerts for matching alert name

Reply:
update the function name